### ISSUE TITLE
Text Builder, ByteString Builder, ShortByteString support in Textual

### ISSUE THREAD

**Issue by TravisCardwell** [ExtremaIS/ttc-haskell#2]
The current design of the `Textual` type class provides support for the following types:

* `String`
* `Text`
* Lazy `Text`
* `ByteString`
* Lazy `ByteString`

Auxiliary functions are required to work with the following types:

Type                   | To Function | From Function
-----------------------|-------------|--------------
`Text` `Builder`       | `TTC.toTLB` | `TTC.fromTLB`
`ByteString` `Builder` | `TTC.toBSB` | `TTC.fromBSB`
`ShortByteString`      | `TTC.toSBS` | `TTC.fromSBS`

The purpose of this issue is to decide whether to add support for the builder and short types to `Textual` or not.

## Options

These options are listed in order of implementation effort, _not_ preference.

### Design A: Current

One option is to keep the current design.

Note that the builder types were included in `Textual` in the initial prototype.  The primary reason for removing them was to decrease the size of the library (number of functions exposed in the API).  I judged that the trade-off is acceptable because auxiliary functions can be used pretty easily.  For reference, please see the initial [design log](https://github.com/ExtremaIS/ttc-haskell/blob/main/project/log/20191123-design.md).

Benefits:

* Fewer functions are exposed in the API than with the other option.

Drawbacks:

* There might exist cases where performance is hurt by passing through a fixed type in order to use the builder types.
* Usage of auxiliary functions is marginally more complex than first-class support.
* Users of the library have to type more to use these types, composing with the auxiliary functions.

### Design B: Inclusion

The other option is to add support for these types to `Textual`.  Doing so is not difficult.

The following functions should be added to the `Textual` type class:

* `toTLB`
* `toBSB`
* `toSBS`

Note that the existing definitions of functions of the same names must be removed.

The following functions should be redefined using `convert`:

* `fromTLB`
* `fromBSB`
* `fromSBS`

The following functions should be added:

* `asTLB`
* `asBSB`
* `asSBS`

Note that these changes cannot be done using a deprecation step.  They would be implemented in a *breaking* major release.  If this design is chosen, I would like to make this change soon, while there are still few users of the library.

Benefits:

* The implementations could make use of all the builder constructors. Conversion to/from a `Text` `Builder` would not need to be done via lazy `Text`, for example.  The alleviates the possible performance concerns.
* With first-class support, it becomes (even more) easy to use these types.

Drawbacks:

* The API gets larger (but only by three functions).

## Notes

Many thanks go to `ephrion` for passionately lamenting the lack of first-class support for builder types in the [TTC announcment](https://old.reddit.com/r/haskell/comments/nr2ext/ann_ttc1000_textual_type_classes/) on the Haskell Reddit.

**Comment by TravisCardwell** [comment]
I am going to move forward with **Design B** (Inclusion).

The `Textual` type class code will get larger, but I do not think there are any significant drawbacks to this.

I will implement Design B and experiment with it.  If it works well, I will merge into the `develop` branch and close this ticket.

**Comment by TravisCardwell** [comment]
This change was a big one, and it added more than three functions to the API:

* The three `to` functions moved into the `Textual` type class.
  * Since this affects all instances, a total of 42 new definitions were required.
* The three `from` functions were reimplemented.
* The three `as` functions were added.
* Three `parse` functions were added.
* Three `parseMaybe` functions were added.
* Three `parseUnsafe` functions were added.

The number of unit tests went from 265 to 470 tests.

The implementation works well, so I merged it into the `develop` branch.

Commit: https://github.com/ExtremaIS/ttc-haskell/commit/2dce892cf40684a1531c175371ca29a2aa89a98c ([Tests](https://github.com/ExtremaIS/ttc-haskell/actions/runs/920386676))
